### PR TITLE
Check if a given PHP Binary is actually valid

### LINKED ISSUE #12: Verify that the PhpBinaryPath created is really a PHP

### PR DESCRIPTION
**Pull request by asgrim**
Fixes #12

### CODE REVIEW

File: src/Platform/TargetPhp/PhpBinaryPath.php [hunk 1]
```diff
@@ -31,7 +33,23 @@ private function __construct(
         public readonly string $phpBinaryPath,
         private readonly string|null $phpConfigPath,
     ) {
-        // @todo https://github.com/php/pie/issues/12 - we could verify that the given $phpBinaryPath really is a PHP install
+        if (! file_exists($this->phpBinaryPath)) {
```
Comment (theofidry) [inline src/Platform/TargetPhp/PhpBinaryPath.php:36]: maybe for testing purposes you may want to be able to create an instance without caring about whether it is valid or not, or for performance reasons in a context where you're sure it is valid. Might be worth having this logic in a static factory? (Could be just `PhpBinaryPath::create()`).

Comment (asgrim) [reply src/Platform/TargetPhp/PhpBinaryPath.php:36]: Yep that seems reasonable, I'll adjust, thank you

**Comment by ghostwriter** [comment]
You are using the wrong directory separator on windows, possibly causing is_executable to “always” return false.

**Comment by asgrim** [comment]
> You are using the wrong directory separator on windows, possibly causing is_executable to “always” return false.

Hmm, looking at this test:

https://github.com/php/php-src/blob/589cfbb24fde06836df9dbb50ee0093a69c6eeba/ext/standard/tests/file/windows_acls/bug44859_3.phpt#L13-L31

The commented-out `tiny.bat` suggests that the `.bat` workaround that I tried won't work. I think it would have to be a true executable on Windows. That suggests the doc:

> On Windows, a file is considered executable, if it is a properly executable file as reported by the Win API GetBinaryType();

is correct, but:

> for BC reasons, files with a .bat or .cmd extension are also considered executable.

is outdated.

**Comment by ghostwriter** [comment]
I’m talking about the path you hardcoded in the test.

image

`/ for Unix-based systems (Linux, macOS).`

`\ for Windows.`

use `DIRECTORY_SEPARATOR` to automatically use the correct one for the platform at runtime.

**Comment by ghostwriter** [comment]
The path: `"z:\phpf\pie\test\unit\Platform\TargetPhp/../../assets/fake-php-windows.bat”`

using both `\` and `/` in the path.


image

**Comment by asgrim** [comment]
> The path: `"z:\phpf\pie\test\unit\Platform\TargetPhp/../../assets/fake-php-windows.bat”`
> 
> using both `\` and `/` in the path.
> 
> 
> image
> 

Yes, I understand what you're saying, but I tried fixing the slashes since my earlier stream, and that is still not the issue. I explored the code some more, and the .bat/.cmd thing does not appear hold true.

**Comment by ghostwriter** [comment]
I see, "normalizing" the path was just the **first** step.

--- 

> … the .bat/.cmd thing does not appear hold true.

you are correct, it does **not** appear hold `true` in this case, but the code does reflect what is documented.

https://github.com/php/php-src/blob/PHP-8.3.11/win32/ioutil.c#L924-L940

--- 

My understanding was, and please correct me If i’m wrong, even if we change the file extension, on windows, if the file’s internal format is **NOT** recognized as executable, `is_executable()` will return `FALSE` (except for file paths with `*.bat` and `*.cmd` extension).

Executability is determined by:

**Windows** returns `TRUE`: 

- If the given file path exists and has a `*.exe` or `*.com` extension, with the [PE format](https://en.wikipedia.org/wiki/Portable_Executable) and internal structure that is recognized as executable using`GetBinaryTypeW`.

**or** 

- If the given file path exists and has a `*.bat` or `*.cmd` extension

**Unix-like Systems** returns `TRUE`: 

- If the given file path exists and the file has the executable bit set.

---

I would not say [outdated](https://github.com/php/php-src/blob/PHP-8.3.11/win32/ioutil.c#L924-L940) but there might be an undocumented behavior/s when we [set the file type and permissions](https://github.com/php/php-src/blob/PHP-8.3.11/win32/ioutil.c#L941-L957) or when [checking executable permission](https://github.com/php/php-src/blob/PHP-8.3.11/win32/ioutil.c#L671-L674) or on files with [empty DACLs](https://github.com/php/php-src/blob/PHP-8.3.11/win32/ioutil.c#L886). 

(empty DACL: no permissions explicitly granted or denied to anyone, effectively making the file inaccessible to all users, including the file's owner.)

---

P.S. Thanks for being patient and indulging my curiosity.

**Pull request merged**